File: web/src/data/remediationModules.ts
/** Remediation modules — must stay aligned with api/app/data/remediation_modules.py */

export type RemediationModuleId =
  | "security_groups"
  | "s3_public_access"
  | "iam_access_keys"
  | "iam_policies"
  | "ssm_parameters"
  | "cloudtrail_logging";

export type RemediationModules = Record<RemediationModuleId, boolean>;

export const DEFAULT_REMEDIATION_MODULES: RemediationModules = {
  security_groups: false,
  s3_public_access: false,
  iam_access_keys: false,
  iam_policies: false,
  ssm_parameters: false,
  cloudtrail_logging: false,
};

export type RemediationModuleSpec = {
  id: RemediationModuleId;
  label: string;
  badgeLabel: string;
  cfnParameter: string;
  summary: string;
  bullets: readonly string[];
  permissions: readonly string[];
  runnerSupported: boolean;
};

export const REMEDIATION_MODULE_SPECS: readonly RemediationModuleSpec[] = [
  {
    id: "security_groups",
    label: "Security Groups",
    badgeLabel: "SG remediation",
    cfnParameter: "EnableSecurityGroupRemediation",
    summary: "Fix open ingress rules",
    bullets: ["Remove open ingress rules", "Restore approved security group configuration"],
    permissions: [
      "ec2:RevokeSecurityGroupIngress",
      "ec2:AuthorizeSecurityGroupIngress",
    ],
    runnerSupported: true,
  },
  {
    id: "s3_public_access",
    label: "S3 public access",
    badgeLabel: "S3 remediation",
    cfnParameter: "EnableS3Remediation",
    summary: "Block public buckets",
    bullets: ["Enforce block public access", "Tighten bucket policies when approved"],
    permissions: ["s3:PutBucketPublicAccessBlock", "s3:PutBucketPolicy"],
    runnerSupported: false,
  },
  {
    id: "iam_access_keys",
    label: "IAM Access Keys",
    badgeLabel: "IAM keys remediation",
    cfnParameter: "EnableIamAccessKeyRemediation",
    summary: "Disable or rotate dormant credentials",
    bullets: ["Disable stale access keys", "Delete unused keys after approval"],
    permissions: ["iam:UpdateAccessKey", "iam:DeleteAccessKey"],
    runnerSupported: false,
  },
  {
    id: "iam_policies",
    label: "IAM policies",
    badgeLabel: "IAM policy remediation",
    cfnParameter: "EnableIamPolicyRemediation",
    summary: "Remove excessive permissions",
    bullets: ["Detach overly broad policies", "Scope inline policies when approved"],
    permissions: [
      "iam:PutRolePolicy",
      "iam:DetachRolePolicy",
      "iam:AttachRolePolicy",
      "iam:CreatePolicyVersion",
      "iam:DeletePolicyVersion",
    ],
    runnerSupported: false,
  },
  {
    id: "ssm_parameters",
    label: "SSM parameters",
    badgeLabel: "SSM remediation",
    cfnParameter: "EnableSsmParameterRemediation",
    summary: "Migrate plaintext secrets",
    bullets: ["Rewrite sensitive String parameters as SecureString", "Run through SSM Automation after approval"],
    permissions: ["ssm:GetParameter", "ssm:PutParameter"],
    runnerSupported: true,
  },
  {
    id: "cloudtrail_logging",
    label: "CloudTrail logging",
    badgeLabel: "CloudTrail remediation",
    cfnParameter: "EnableCloudTrailRemediation",
    summary: "Enable logging if disabled",
    bullets: ["Start logging on trails", "Update trail configuration when approved"],
    permissions: ["cloudtrail:UpdateTrail", "cloudtrail:StartLogging"],
    runnerSupported: false,
  },
];

export function anyRemediationEnabled(modules: RemediationModules): boolean {
  return REMEDIATION_MODULE_SPECS.some((m) => modules[m.id]);
}

export function countRemediationEnabled(modules: RemediationModules): number {
  return REMEDIATION_MODULE_SPECS.filter((m) => modules[m.id]).length;
}

File: api/app/data/remediation_modules.py
"""Remediation automation modules — scoped write per finding family."""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True)
class RemediationModuleSpec:
    id: str
    label: str
    badge_label: str
    enable_column: str
    deployed_column: str
    cfn_parameter: str
    iam_policy_name: str
    permissions: tuple[str, ...]
    runner_supported: bool


REMEDIATION_MODULES: tuple[RemediationModuleSpec, ...] = (
    RemediationModuleSpec(
        id="security_groups",
        label="Security Groups",
        badge_label="SG remediation",
        enable_column="enable_remediation_sg",
        deployed_column="remediation_sg_deployed",
        cfn_parameter="EnableSecurityGroupRemediation",
        iam_policy_name="VigilRemediationSecurityGroups",
        permissions=(
            "ec2:RevokeSecurityGroupIngress",
            "ec2:AuthorizeSecurityGroupIngress",
        ),
        runner_supported=True,
    ),
    RemediationModuleSpec(
        id="s3_public_access",
        label="S3 public access",
        badge_label="S3 remediation",
        enable_column="enable_remediation_s3",
        deployed_column="remediation_s3_deployed",
        cfn_parameter="EnableS3Remediation",
        iam_policy_name="VigilRemediationS3",
        permissions=(
            "s3:PutBucketPublicAccessBlock",
            "s3:PutBucketPolicy",
        ),
        runner_supported=False,
    ),
    RemediationModuleSpec(
        id="iam_access_keys",
        label="IAM access keys",
        badge_label="IAM keys remediation",
        enable_column="enable_remediation_iam_keys",
        deployed_column="remediation_iam_keys_deployed",
        cfn_parameter="EnableIamAccessKeyRemediation",
        iam_policy_name="VigilRemediationIamAccessKeys",
        permissions=(
            "iam:UpdateAccessKey",
            "iam:DeleteAccessKey",
        ),
        runner_supported=False,
    ),
    RemediationModuleSpec(
        id="iam_policies",
        label="IAM policies",
        badge_label="IAM policy remediation",
        enable_column="enable_remediation_iam_policy",
        deployed_column="remediation_iam_policy_deployed",
        cfn_parameter="EnableIamPolicyRemediation",
        iam_policy_name="VigilRemediationIamPolicy",
        permissions=(
            "iam:PutRolePolicy",
            "iam:DetachRolePolicy",
            "iam:AttachRolePolicy",
            "iam:CreatePolicyVersion",
            "iam:DeletePolicyVersion",
        ),
        runner_supported=False,
    ),
    RemediationModuleSpec(
        id="ssm_parameters",
        label="SSM parameters",
        badge_label="SSM remediation",
        enable_column="enable_remediation_ssm_parameters",
        deployed_column="remediation_ssm_parameters_deployed",
        cfn_parameter="EnableSsmParameterRemediation",
        iam_policy_name="VigilRemediationAutomation",
        permissions=(
            "ssm:GetParameter",
            "ssm:PutParameter",
        ),
        runner_supported=True,
    ),
    RemediationModuleSpec(
        id="cloudtrail_logging",
        label="CloudTrail logging",
        badge_label="CloudTrail remediation",
        enable_column="enable_remediation_cloudtrail",
        deployed_column="remediation_cloudtrail_deployed",
        cfn_parameter="EnableCloudTrailRemediation",
        iam_policy_name="VigilRemediationCloudTrail",
        permissions=(
            "cloudtrail:UpdateTrail",
            "cloudtrail:StartLogging",
        ),
        runner_supported=False,
    ),
)

REMEDIATION_MODULE_BY_ID = {m.id: m for m in REMEDIATION_MODULES}
DEFAULT_REMEDIATION_ROLE_NAME = "VigilRemediationAutomationRole"


def remediation_modules_dict(acc: Any) -> dict[str, bool]:
    return {m.id: bool(getattr(acc, m.enable_column)) for m in REMEDIATION_MODULES}


def remediation_deployed_dict(acc: Any) -> dict[str, bool]:
    return {m.id: bool(getattr(acc, m.deployed_column)) for m in REMEDIATION_MODULES}


def any_remediation_enabled(modules: dict[str, bool]) -> bool:
    return any(modules.values())


def set_remediation_modules(acc: Any, modules: dict[str, bool]) -> None:
    for spec in REMEDIATION_MODULES:
        setattr(acc, spec.enable_column, bool(modules.get(spec.id, False)))


def clear_remediation_deployed(acc: Any, *, module_id: str | None = None) -> None:
    if module_id:
        spec = REMEDIATION_MODULE_BY_ID[module_id]
        setattr(acc, spec.deployed_column, False)
        return
    for spec in REMEDIATION_MODULES:
        setattr(acc, spec.deployed_column, False)
